text/0000-per-project-config.md

```diff
@@ -0,0 +1,87 @@
+
+This will likely require that the `Project` and `Hooks` objects in `notion-core` are merged in some way into a single configuration object that pulls from all possible sources.
+
+## `notion config` Reference
+
+To help users understand which configurations are active and where they are coming from, we should implement a `notion config list` command that works similarly to `npm config list`. This will allow users to immediately see what Notion understands about the state of the world from a given place in the filesystem, while also providing us with a useful tool for debugging the implementation. This also opens the option going forward of providing CLI commands for editing the configuration (e.g. `notion config set`).
+
+## Multi-stage Implementation
+
+To facilitate the work and allow users to work with parts of these changes before we make an entire overhaul to the system, we can implement this RFC in multiple stages:
+
+1. Add support for parsing a project-local `notion-hooks.toml` file, if it exists, as well as the `~/.notion/hooks.toml` that we currently check for.
+2. Rename `hooks.toml` into `.notionrc.json`, changing the file format but supporting the same set of configurations (i.e. Only hooks, no platform specifications at this point).
+3. Fully merge the hooks config with the platform config internally and add support for `toolchain` within `.notionrc.json`
+
+# Critique
+[critique]: #critique
+
+## Configuration Precedence
+
+One possible update is that a different precedence order could be used for the configurations. However, it is a common pattern in computing that more specific settings beat less specific settings, so we should still adhere to that. Additionally, the specific-to-general precedence order matches the order used by `npm` for `npmrc` files.
+
+As discussed above, the project-level `.notionrc.json` setting is considered to be a more "advanced" way of specifying per-project settings, so it should have precedence over `package.json`.
+
+## Merging Configurations
+
+Another approach to determining the final configuration from the available files would be to use _only_ the highest-precedence `.notionrc.json` if it is present and fall back to the existing `package.json` if there isn't a `.notionrc.json` present. This would mean that if any `.notionrc.json` is present, it represents the single source of truth for all of the Notion settings. However, it would remove the ability to have settings cascade down from the user-level settings to the project-level settings when a project doesn't have any custom configuration beyond what the user has specified.
+
+The merge approach suggested in the RFC matches how `npm` resolves configuration from multiple sources, pulling each value in precedence order, but allowing different files to specify different settings.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- What additional settings, if any, should we include in `.notionrc.json`?
```

Comment:
I think this is out of scope for this RFC, but in the future we might want to add config for a version specifier for `notion`? In case the project uses some Notion feature that is only available after a certain release?

Reply:
Definitely, that's a great idea and will allow a nice, actionable error message early in the process, instead of some opaque one later when something fails. Or even worse, doing the wrong thing because we don't even read a new configuration at all.